### PR TITLE
topdev531 // request changes for participants

### PR DESCRIPTION
**Pull request by ortegacmanuel**
@topdev531 this is your branch with your changes.

### CODE REVIEW

**Comment by ortegacmanuel** [comment]
@topdev531 We had to remove your changes from the `starting_point` branch so other can use it as an starting point.

Sorry about that, our bad, we didn't protected the starting_point branch.

Thanks!

**Comment by jmetzner9** [comment]
## Description

This update contains:

- Migration to add new column status to participants table
- Update for changes: create participant and update status
- Using RSpec for unit testing

Here my assumption: the email is uniq, the status only can be updated if it is pending.

Some can be improved:

- Using aasm gem to manage status transition
- Split update status logic into a service class

## How Has This Been Tested?

`bundle install`
`rails db:migrate`
`rails db:migrate RAILS_ENV=test`
`rspec -fd spec/controllers/api/v1/participants_controller_spec.rb`

**Comment by topdev531** [comment]
> starting point.
> 
> Sorry about that, our bad, we didn't protected the starting_point branch.
> 
> Thanks!

**Comment by jmetzner9** [comment]
@ortegacmanuel  Sorry, I was checking my pull request again and clicked Close accidentally, so I reopened it again.

File: app/controllers/api/v1/participants_controller.rb [hunk 1]
```diff
@@ -2,7 +2,42 @@ module Api::V1
   class ParticipantsController < ApplicationController
 
     def index
-      render json: Participant.all
+      render json: Participant.accepted
```
Comment (ortegacmanuel) [inline app/controllers/api/v1/participants_controller.rb:6]: @topdev531 great work adding an scope!

File: app/controllers/api/v1/participants_controller.rb [hunk 2]
```diff
@@ -2,7 +2,42 @@ module Api::V1
...
+    end
+
+    def create
+      participant = Participant.new(participant_params)
```
Comment (ortegacmanuel) [inline app/controllers/api/v1/participants_controller.rb:10]: strong params :+1:

File: app/controllers/api/v1/participants_controller.rb [hunk 3]
```diff
@@ -2,7 +2,42 @@ module Api::V1
...
+
+      if participant.save
+        render json: participant
+      else
+        render json: { error: participant.errors.full_messages }
+      end
+    end
+
+    def update
+      unless Participant.statuses.keys.include?(params[:status])
```
Comment (ortegacmanuel) [inline app/controllers/api/v1/participants_controller.rb]: nice constraint! 

I think we could make this more readable doing something like this

```
  def update
     ....
     return invalid_status unless status_param_valid?
     ....
  end

  private

  def invalid_status
    render json: { message: "Invalid status" }
  end

  def status_param_valid?
    Participant.statuses.keys.include?(params[:status])
  end
```

just saying! :+1:  @topdev531

File: app/controllers/api/v1/participants_controller.rb [hunk 4]
```diff
@@ -2,7 +2,42 @@ module Api::V1
...
+        render json: { error: 'Invalid status' }
+        return
+      end
+
+      if participant = Participant.find_by_id(params[:id])
```
Comment (ortegacmanuel) [inline app/controllers/api/v1/participants_controller.rb]: @topdev531  here we could extract the handling of the ActiveRecord::RecordNotFound exception to the controller class

```
  rescue_from ActiveRecord::RecordNotFound, with: :not_found

  def not_found
    render json: { message: "Not Found" }, status: 404
  end
```

And in the action we will only need 
```
        participant = Participant.find(params[:id])
        if participant.pending?
          participant.status = params[:status]
          participant.save
          render json: participant
        else
          render json: { error: "Participant has already been #{participant.status}" }
        end

```

File: spec/controllers/api/v1/participants_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
+require 'rails_helper'
```
Comment (ortegacmanuel) [inline spec/controllers/api/v1/participants_controller_spec.rb:1]: Nice work adding test coverage!

**Comment by ortegacmanuel** [comment]
@topdev531  great work! thanks for sending in a submission. We just added a review. No rush or need to do the changes. Adding it we only want to share with you some thoughts on the code and show you how we normally work/review PRs. 

we'll get in contact soon. Thanks!

**Comment by jmetzner9** [comment]
@ortegacmanuel  Thanks for reviewing my work and your comments are really helpful for improving my skills, I will keep your advice on future work and it will be a great experience for me.

Will wait for your reply.

**Pull request closed without merging**